Composer/packages/server/src/models/connector/csharpBotConnector.ts

```diff
@@ -2,74 +2,178 @@
+      configList.push('--MicrosoftAppPassword');
+      configList.push(config.MicrosoftAppPassword);
+    }
+    if (config.luis) {
+      if (config.luis.authoringKey) {
+        configList.push('--luis:endpointKey');
+        configList.push(config.luis.authoringKey);
+      }
+      if (config.luis.authoringRegion) {
+        configList.push('--luis:endpoint');
+        configList.push(`https://${config.luis.authoringRegion}.api.cognitive.microsoft.com`);
+      }
     }
 
-    this.status = BotStatus.NotConnected;
+    return configList;
+  };
+  private addListeners = (child: ChildProcess, handler: Function) => {
+    if (child.stdout !== null) {
+      child.stdout.on('data', (data: any) => {
+        log(`stdout: ${data}`);
+      });
+    }
+
+    if (child.stderr !== null) {
+      child.stderr.on('data', (data: any) => {
+        log(`stderr: ${data}`);
+      });
+    }
+
+    child.on('close', code => {
+      log(`close ${code}`);
+      handler();
+    });
+
+    child.on('error', (err: any) => {
+      log(`stderr: ${err}`);
+    });
+
+    child.on('exit', code => {
+      log(`exit: ${code}`);
+      handler();
+    });
 
-    return `${this.endpoint}/api/messages`;
+    child.on('message', msg => {
+      log(msg);
+    });
+
+    child.on('disconnect', code => {
+      log(`disconnect: ${code}`);
+      handler();
+    });
   };
 
-  sync = async (config: DialogSetting) => {
-    // archive the project
-    // send to bot runtime service
+  private getBotPath = () => {
     const currentProject = BotProjectService.getCurrentBotProject();
     if (currentProject === undefined) {
       throw new Error('no project is opened, nothing to sync');
     }
-    const dir = Path.join(currentProject.dataDir);
-    const luisConfig = currentProject.luPublisher.getLuisConfig();
-    await this.archiveDirectory(dir, './tmp.zip');
-    const content = fs.readFileSync('./tmp.zip');
+    return Path.join(currentProject.dir);
+  };
 
-    const form = new FormData();
-    form.append('file', content, 'bot.zip');
+  private start = async (dir: string, config: DialogSetting) => {
+    const runtime = spawn(
+      'dotnet',
+      ['bin/Debug/netcoreapp2.1/BotProject.dll', `--urls`, this.endpoint, ...this.getConnectorConfig(config)],
+      {
+        detached: true,
+        cwd: dir,
+        stdio: ['ignore', 'ignore', 'inherit'],
+      }
+    );
+    log(`runtime started. pid: ${runtime.pid}`);
+    this.addListeners(runtime, this.stop);
+    CSharpBotConnector.botRuntimes[runtime.pid] = runtime;
+    this.status = BotStatus.Connected;
+  };
 
-    if (luisConfig && luisConfig.authoringKey !== null && !currentProject.checkLuisPublished()) {
-      throw new Error('Please publish your Luis models');
-    }
+  private checkPortUsable = async (port: string | number): Promise<string> => {
+    return new Promise((resolve, reject) => {
+      const netstat = spawn('netstat', ['-ano']);
+      const findstr = spawn('findstr', [`${port}`], {
```

Comment:
`findstr` does not exist on my machine (a mac).

Reply:
I use lsof to check the port availability in not win32 platform.